design/secrets.md

```diff
@@ -0,0 +1,81 @@
+      If the second option, the new flag will be `--credentials string` where the argument will be the name of the credential set when adding a plugin.
+    - The `velero snapshot-location (create|set)` will have a new flag to set the credentials based on the option selected above. If the first option, the new flag will be `--credentials mapStringString` which sets the name of the corresponding credentials secret for a provider. Format is provider:credentials-secret-name.
+      If the second option, the new flag will be `--credentials string` where the argument will be the name of the credential set when adding a plugin.
+
+  Note that for this logic to work we must have a controller loop checking for when a corresponding secret is present before marking the BSL/VSL as ready.
+
+- The spec for BSLs and VSLs will updated to contain a new field `Credentials`. The value in this field will be the name of the key in the `cloud-credentials` secret to use for authenticating with the storage provider.
+
+- Plugins will need to be invoked differently so that the correct credential is used.
+  Currently, there is a single secret, which is mounted into every pod deployed by Velero (the Velero Deployment and the Restic DaemonSet).
+  This secret contains a single data key (`cloud`) which is accessible within the pods at the path `/credentials/cloud`.
+  This path is made known to all plugins through provider specific environment variables.
+  All possible provider environment variables are set to this path.
+  Instead of setting the environment within all the pods, we can modify `restartableProcess` to set the environment variables before running a plugin process.
+  Each plugin process would still have the same set of environment variables set, however the value used for each of these variables would instead be a different path within the `/credentials` directory, formed from the credential selected for a particular BSL/VSL.
+  Taking this approach would not require any changes from plugins as the credentials information would be made available to them in the same way.
+  We will also need to ensure that the restic controllers are updated in the same way so that correct credentials are used (when creating a `ResticRepository` or processing `PodVolumeBackup`/`PodVolumeRestore`).
+
+## Alternatives Considered
+
+Instead of associating credentials with plugins, credentials will exist as standalone entities which can be queried using a new command `velero credentials`.
+This will be a wrapper around Kubernetes API calls to query the secret used by Velero.
+Credentials can be listed using `velero credentials get`, which will list the names of the keys within the `cloud-credentials` secret.
+Credentials can be added or removed using `velero credentials add|delete`.
+
+The rest of the workflow of how to use these credentials would still apply, such as specifying the name of the credential to use for a particular BSL/VSL.
+
+## Security Considerations
+
+N/A
```

Comment:
Given we're working with secrets, I think security applies here, and we should write out what's happening, even if the exposed data ends up being the same.